Doc for Memfault Core Metrics (#703)
memfault/memfault-docs: docs/platform/memfault-core-metrics.mdx

---
title: Memfault Core Metrics
---

import Tabs from "@theme/Tabs";
import TabItem from "@theme/TabItem";

> Memfault provides a set of common built-in metrics that apply to Android, Linux,

Reviewer: I'd strongly consider removing the part where we say that they apply to "Android, Linux and MCU". I guess as a customer, I'm not sure if I'd care that much if metrics applied to multiple different devices? versus that the metrics just made sense for my platform. 

extra nit, I'm not sure "Memfault provides" is the right framing, maybe either "Memfault identified a set of useful universal metrics that apply..."

I'd also remove "built-in" here because it's a bit weird to say here and then repeat the collection aspect in the next sentence - if we're keeping the "provides" framing, I'd probably merge the 1st and 2nd sentences.

Author: good call, i switched the leading sentence to "Memfault provides support for a set of core metrics that apply to a wide range of devices."

and MCU devices. These metrics are either automatically collected, or have
built-in facilities for easing collection. Memfault's server will automatically
process these metrics to provide insights on a few key performance indicators
for the fleet. Memfault considers these metrics to be relevant to a wide range
of devices.

The Memfault Core Metrics share the following properties:

- either automatically collected or have built-in facilities for easing
  collection
- exempt from Custom Metric/Attribute quotas

> - automatic insights in Memfault's dashboard (coming soon!)

Reviewer: How do we feel about not having these leading paragraphs, and just directly talking about the metrics? 

I'm thinking about Sentry's docs where they just ignore the preamble and go into things, and it feels a bit more like, trust us, we know what we're doing?

https://docs.sentry.io/platforms/android/performance/instrumentation/automatic-instrumentation/

I don't necessarily love a public facing name of "Core Metrics" since it's a bit presumptious for us to call them Core Metrics for others. I assume there's already been discussions about naming so I'm probably rehashing stuff :d

Reviewer 2: I would take an alternate standpoint and say maybe what this document needs is a good explanation of why these metrics are important and useful (I think you took a stab at this internally @noahp). Just linking to external blogpost and docs is a little weak and maybe this should be the new place where we say why they are important.

Author: good point- I touched it up to include a short paragraph explaining the "why" for each metric. @Blisse I kinda picked the name in haste, so if you have a better one I'd love to swap it in :D

The appropriate SDK needs to be integrated before these metrics can be used. See
the guides linked from each platform's section for getting up and running on the

> Memfault platform.

Reviewer: We should probably linked which version of the Memfault SDK these metrics first appear on 

4.10.0 for Bort, I believe

Reviewer 2: ```suggestion
Memfault platform.

| Platform | Required SDK Version |
| -------- | -------------------- |
| Android  | > 4.11               |
| Linux    | > 1.9                |
| MCU      | > XX                 |
```

A summary of the Memfault Core Metrics provided by the device SDKs:

| Metric Key                    | Automatically Collected on these SDKs | Category                                                  |
| ----------------------------- | ------------------------------------- | --------------------------------------------------------- |

> | battery_soc_pct_drop          | Android, Linux                        | [Battery Life](#battery)                                  |
> | battery_discharge_duration_ms | Android, Linux                        | [Battery Life](#battery)                                  |
> | battery_soc_pct               | Android, Linux                        | [Battery Life](#battery)                                  |

Reviewer 2: Not automatic on Linux either.

```suggestion
| battery_soc_pct_drop          | Android                        | [Battery Life](#battery)                                  |
| battery_discharge_duration_ms | Android                        | [Battery Life](#battery)                                  |
| battery_soc_pct               | Android                        | [Battery Life](#battery)                                  |
```

| operational_hours             | Android, Linux, MCU                   | [Crash-Free Hours](#crash-free)                           |
| operational_crashfree_hours   | Android, Linux, MCU                   | [Crash-Free Hours](#crash-free)                           |
| sync_successful               |                                       | [Session Based Connectivity](#session-based-connectivity) |
| sync_failure                  |                                       | [Session Based Connectivity](#session-based-connectivity) |
| sync_memfault_successful      | Android, Linux                        | [Session Based Connectivity](#session-based-connectivity) |
| sync_memfault_failure         | Android, Linux                        | [Session Based Connectivity](#session-based-connectivity) |
| connectivity_connected_time   |                                       | [Always On Connectivity](#always-on-connectivity)         |

> | connectivity_expected_time    |                                       | [Always On Connectivity](#always-on-connectivity)         |

Reviewer: These are automatically collected on Android - though we might be having some conversations about whether they're 100% accurate..

## Battery Life {#battery}

For background and principles around these metrics, see the following docs:

- [Documentation of best practices for battery life tracking](/docs/best-practices/metrics-for-battery-life)
- [Debugging Android Battery Life with Memfault](/docs/best-practices/android-battery-debugging)

The Battery Life metrics are:

- `battery_soc_pct_drop`: The drop of the battery's State of Charge (SoC) during
  the heartbeat interval, reported as a percentage from 1-100.
- `battery_discharge_duration_ms`: The time spent discharging the battery during
  the heartbeat interval.
- `battery_soc_pct`: The battery's State of Charge (SoC) at the end of the
  heartbeat interval.

Battery State of Charge metrics collection depends on the platform.

<Tabs queryString="battery-soc-platform">

>   <TabItem value="Android + Linux">
> 
> Battery metrics are collected automatically on devices running the Memfault
> Android or Linux SDKs.
> 
>   </TabItem>

Reviewer 2: ```suggestion
  <TabItem value="Android">

Battery metrics are collected automatically on devices running the Memfault
Android SDK.

  </TabItem>
  <TabItem value="Linux">Coming soon!</TabItem>

```

[nRF-Connect SDK](/docs/mcu/nordic-nrf-connect-sdk-guide)) and the
[ESP-IDF SDK](/docs/mcu/espressif-esp32-guide); other platforms can manually
record those metrics by enabling the API and calling it directly when syncing
Memfault data:

```c title="memfault_platform_config.h
// Enable the sync_memfault_* metrics
#define MEMFAULT_METRICS_MEMFAULT_SYNC_SUCCESS 1
```

```c reference title="memfault/metrics/connectivity.h" referenceLinkText="See header file on GitHub"
https://github.com/memfault/memfault-firmware-sdk/blob/d23835cb1144d9ca3a217be9d972e4f8dc3e165a/components/include/memfault/metrics/connectivity.h#L24-L29
```

:::

### Always On Connectivity {#always-on-connectivity}

Devices with continuous connectivity can use these metrics to track the relative
uptime of their connectivity:

- `connectivity_connected_time_ms`: The time spent connected, in milliseconds
- `connectivity_expected_time_ms`: The time expected to be connected, in
  milliseconds

Examples of devices with continuous connectivity include:

- Mains powered WiFi connected devices
- Cellular connected devices
- Ethernet connected devices
- Mains powered Zigbee/802.15.4/Thread devices

> This connectivity metric is used by Memfault to compute relative uptime.

Reviewer: Maybe add this line after again:

```
Always On Connectivity Metrics collection depends on the platform.
```

(it's automatically collected by Android
